src/main/java/backend/RepoIO.java

```diff
@@ -51,14 +52,32 @@ public RepoIO(boolean isTestMode, boolean enableTestJSON) {
 
     public CompletableFuture<Model> openRepository(String repoId) {
         if (repoStore.isRepoStored(repoId)) {
-            return repoStore.loadRepository(repoId)
-                .thenCompose(this::updateModel)
-                .exceptionally(withResult(new Model(repoId)));
+            return loadRepositoryFromStore(repoId);
         } else {
-            return repoSource.downloadRepository(repoId)
+            return downloadRepositoryFromSource(repoId);
+        }
+    }
+
+    private CompletableFuture<Model> loadRepositoryFromStore(String repoId) {
+        return repoStore.loadRepository(repoId)
+                .thenCompose(this::updateModel)
+                .exceptionally(ex -> {
+                    try {
```

Comment:
An alternative way to see why it won't compile is that `Runnable::run` is not declared as throwing any exceptions, and `ExecutionException` and `InterruptedException` are both checked.

The try/catch is unfortunately necessary, but it'd be nice to factor this out, possibly into something more generic. Maybe a new function, `withFuture`, so we can do something like this?

```java
return repoStore.loadRepository(repoId)
    .thenCompose(this::updateModel)
    .exceptionally(withFuture(downloadRepositoryFromSource(repoId)));
```
This way we can reuse existing error-handling from `downloadRepositoryFromSource`.

The implementation should be fairly similar to [withResult](https://github.com/HubTurbo/HubTurbo/blob/master/src/main/java/util/Futures.java#L60).